external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala

```diff
@@ -0,0 +1,199 @@
+    * or greater than the specified object.
+    */
+  override def compareTo(other: Offset): Int = other match {
+    case KafkaSourceOffset(otherOffsets) =>
+      val allTopicAndPartitions = (this.offsets.keySet ++ otherOffsets.keySet).toSeq
+
+      val comparisons = allTopicAndPartitions.map { tp =>
+        (this.offsets.get(tp), otherOffsets.get(tp)) match {
+          case (Some(a), Some(b)) =>
+            if (a < b) {
+              -1
+            } else if (a > b) {
+              1
+            } else {
+              0
+            }
+          case (None, _) =>
+            -1
+          case (_, None) =>
+            1
+          case _ =>
+            throw new IllegalArgumentException(
+              s"Invalid comparison between two sets of Kafka offsets: $this <=> $other")
+        }
+      }
+      val nonZeroSigns = comparisons.filter { _ != 0 }.toSet
+      nonZeroSigns.size match {
+        case 0 => 0 // if both empty or only 0s
+        case 1 => nonZeroSigns.head // if there are only (0s and 1s) or (0s and -1s)
+        case _ => // there are both 1s and -1s
+          throw new IllegalArgumentException(
+            s"Invalid comparison between non-linear histories: $this <=> $other")
+      }
+
+    case _ =>
+      throw new IllegalArgumentException(s"Cannot compare $this <=> $other")
+  }
+
+  override def ==(other: Offset): Boolean = Try(compareTo(other) == 0).getOrElse(false)
+
+  /** Returns a set of offset ranges between `this` and `other` */
+  def to(other: KafkaSourceOffset): Seq[OffsetRange] = {
+
+    // Get all the partitions referenced in both sets of offsets
+    val allTopicAndPartitions = (this.offsets.keySet ++ other.offsets.keySet).toSeq
+
+    // For each partition, figure out the non-empty ranges of offsets
+    allTopicAndPartitions.flatMap { tp =>
+      (this.offsets.get(tp), other.offsets.get(tp)) match {
+
+        // Data was read till fromOffset and needs to be read till untilOffset
+        case (Some(fromOffset), Some(untilOffset)) =>
+          if (untilOffset > fromOffset) {
+            Some(OffsetRange(tp, fromOffset, untilOffset))
+          } else None
+
+        // TODO: Support cases where topic+partitions are missing from one. Can happen in case of
+        // repartitioning.
+
+        case _ =>
+          None
+      }
+    }
+  }
+
+  override def toString(): String = {
+    offsets.toSeq.sortBy(_._1.topic).mkString("[", ", ", "]")
+  }
+}
+
+/** Companion object of the [[KafkaSourceOffset]] */
+private[kafka] object KafkaSourceOffset {
+
+  /** Returns [[KafkaSourceOffset]] from a Option[Offset]. */
+  def from(offsetOption: Option[Offset]): Option[KafkaSourceOffset] = {
+    offsetOption.map { offset =>
+      offset match {
+        case o: KafkaSourceOffset => o
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Invalid conversion from offset of ${offset.getClass} to $getClass")
```

Comment:
This message is a little misleading since `getClass` will the return the name of the singleton.